privilege/privileges/tidb_auth_token.go

```diff
@@ -0,0 +1,110 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package privileges
+
+import (
+	"context"
+	"encoding/json"
+	"strings"
+	"sync"
+	"time"
+
+	jwkRepo "github.com/lestrrat-go/jwx/v2/jwk"
+	jwsRepo "github.com/lestrrat-go/jwx/v2/jws"
+	jwtRepo "github.com/lestrrat-go/jwx/v2/jwt"
+	"github.com/pingcap/errors"
+	"github.com/pingcap/tidb/util/logutil"
+	"go.uber.org/zap"
+)
+
+// JWKSImpl contains a JSON Web Key Set (JWKS), and a filepath that stores the JWKS
+type JWKSImpl struct {
+	// contain one more JWKS for recover when failing to load the JWKS from the filepath.
+	cur, last jwkRepo.Set
+	m         sync.RWMutex
+	filepath  string
+}
+
+// GlobalJWKS is the global JWKS for tidb-server
+var GlobalJWKS JWKSImpl
+
+func (jwks *JWKSImpl) load() (err error) {
+	jwks.m.Lock()
+	defer jwks.m.Unlock()
+	jwks.last = jwks.cur
+	jwks.cur, err = jwkRepo.ReadFile(jwks.filepath)
+	if err != nil {
+		jwks.cur = jwks.last
+	}
+	return err
+}
+
+func (jwks *JWKSImpl) verify(tokenBytes []byte) (payload []byte, err error) {
+	jwks.m.RLock()
+	defer jwks.m.RUnlock()
+	return jwsRepo.Verify(tokenBytes, jwsRepo.WithKeySet(jwks.cur))
+}
+
+// LoadJWKS4AuthToken reload the jwks every auth-token-refresh-interval.
+func (jwks *JWKSImpl) LoadJWKS4AuthToken(jwksPath string, interval time.Duration, reload bool) error {
+	jwks.filepath = jwksPath
+	if reload {
+		go func() {
```

Comment:
It's better to manage the goroutine by `waitgroup.WaitGroup` and close the goroutine when the server shuts down.

Reply:
Since there is only one goroutine performing reloading, I prefer not to add `WaitGroup` until we need to support changing the path of JWKS (and therefore call `LoadJWKS4AuthToken` again).

Reply:
It is not about reloading or dynamic path. It is about the exit of the goroutine.